documentation/ir.md

```diff
@@ -0,0 +1,694 @@
+```
+
+<!-- TOC --><a name="qualification"></a>
+### Qualification
+
+> [!IMPORTANT]
+> TODO: FINISH THIS SECTION.
+
+<!-- TOC --><a name="hybrid-tree"></a>
+## Hybrid Tree
+
+> [!IMPORTANT]
+> TODO: FINISH THIS SECTION.
+
+For an example of the Hybrid Tree structure, consider the `nation_info` example earlier from the unqualified nodes:
+
+```py
+nation_info = nations.CALCULATE(
+  region_name=region.name,
+  nation_name=name,
+  n_customers_in_debt=COUNT(customers.WHERE(acctbal < 0).orders)
+).ORDER_BY(nation_name.ASC())
+```
+
+```mermaid
+flowchart TD
+  subgraph L1["(H1)"]
+  A[Root]
+  end
+  subgraph L2["(H2)"]
+  direction LR
+    B[Collection:
+    nations]
+    C["Calculate
+    region_name: $0.name
+    n_customers_in_debt: DEFAULT_TO($1.agg_0, 0)
+    nation_name: name
+    "]
+    D["OrderBy
+    nation_name (ascending)"]
+    B --> C --> D
+  end
+  subgraph L4c["($1)"]
+  subgraph L4["(H4)"]
+    direction LR
+    R[Sub-Collection:
+    customers]
+    S[Where:
+    acctbal < 0]
+    R --> S
+  end
+  subgraph L5["(H5)"]
+    direction LR
+    V[Sub-Collection:
+    orders]
+  end
+  T["Child Info:
+  join_keys: [(key, nation_key)]
+  agg_keys: [nation_key]
+  aggs: {'agg_0': COUNT()}
+  "]
+  end
+  subgraph L3c["($0)"]
+  subgraph L3["(H3)"]
+  Q[Sub-Colection:
+  region]
+  end
+  U["Child Info:
+  join_keys: [(region_key, key)]
+  "]
+  end
+  L1 --> L2
+  L2 -.->|Singular| L3c
+  L2 -.->|Aggregation| L4c
+  L4 --> L5
+```
+
+In this example, the main hybrid tree has two levels `H1` and `H2` (`H2` is the value pointed to when the QDAG is converted to a hybrid tree).
+- `H1`/`H2` have a parent/successor relationship where `H1` is the parent and `H2` is the successor.
+- `H1` has no children and its pipeline has a single operation denoting the root level.
+- `H2` has two children (`$0` and `$1`) and its pipeline has 3 operations:
+  - An access to the nations collection (how the step-down from `H1` to `H2` begins)
+  - A calculate that defines `region_name`, `nation_name`, and `n_customers_in_debt`
+  - An order-by that sorts by `nation_name` in ascending order.
+- `$0` is a singular access, meaning the data from `H2` and `$0` can be directly joined without needing to worry about changes in cardinality. They are joined on the condition that the `region_key` term from `H2` equals the `key` term from the bottom subtree of `$0` (which is `H3`).
+  - The contents of `$0` is just a single tree level `H3` which does not have any children and has a pipeline containing only an access to the `regions` collection.
+- `$1` is an aggregation access, meaning the data from `$1` must first be aggregated before it is joined with `H2`. The aggregation is done by grouping on the `nation_key` field of the bottom subtree of `$1` (which is `H5`) and computes the term `agg_0` as `COUNT()`. Then, the result is joined with `H2` on the condition that the `key` term from `H2` equals the `nation_key` field just used to aggregate.
```

Comment:
Suggestion to make $0 and $1 points have sublists describing steps rather than paragraph.
E.g.:
$1 is an aggregation access over customers.orders, with:
Grouped by nation_key and joined to H2 on key = nation_key
COUNT() aggregation over the orders collection
Join on condition ....